book/src/plumbing/db_lifetime.md

```diff
@@ -0,0 +1,238 @@
+This defining query is unique within a revision, meaning that `f` executes at most once with that same set of arguments.
+
+We say that a query has *executed in a revision R* if its function body was executed. When this occurs, all tracked structs defined (created) by that query will be recorded along with the query's result.
+
+We say that a query has been *validated in a revision R* if the salsa system determined that its inputs did not change and so skipped executing it. This also triggers the tracked structs defined by that query to be considered validated (in particular, we execute a function on them which updates some internal fields, as described below).
+
+When we talk about `ts`, we mean 
+
+## Theorem: At the start of a new revision, all references to `ts` are within salsa's database
+
+After `ts` is deleted, there may be other memoized values still reference `ts`, but they must have a red input query.
+**Is this true even if there are user bugs like non-deterministic functions?**
+Argument: yes, because of non-forgery, those memoized values could not be accessed.
+How did those memoized values obtain the `TS<'db>` value in the first place?
+It must have come from a function argument (XX: what about thread-local state).
+Therefore, to access the value, they would have to provide those function arguments again.
+But how did they get them?
+
+Potential holes:
+
+* Thread-local APIs that let you thread `'db` values down in an "invisible" way, so that you can return them without them showing up in your arguments -- e.g. a tracked function `() -> S<'db>` that obtains its value from thread-local state.
+    * We might be able to sanity check against this with enough effort by defining some traits that guarantee that every lifetime tagged thing in your result *could have* come from one of your arguments, but I don't think we can prove it altogether. We either have to tell users "don't do that" or we need to have some kind of dynamic check, e.g. with a kind of versioned pointer. Note that it does require unsafe code at present but only because of the limits of our existing APIs.
+    * Alternatively we can do a better job cleaning up deleted stuff. This we could do.
+* what about weird `Eq` implementations and the like? Do we have to make those unsafe?
+
+## Theorem: To access a tracked struct `ts` in revision R, the defining query `f(..)` must have either *executed* or been *validated* in the revision R.
+
+This is the core bit of reasoning underlying most of what follows.
+The idea is that users cannot "forge" a tracked struct instance `ts`.
+They must have gotten it through salsa's internal mechanisms.
+This is important because salsa will provide `&`-references to fields within that remain valid during a revision.
+But at the start of a new revision salsa may opt to modify those fields or even free the allocation.
+This is safe because users cannot have references to `ts` at the start of a new revision.
```

Comment:
Nit: I'm not sure if that's mentioned above. But I think that's because the operations require a `&'db mut Db`

Reply:
Yes